File: 744-network-delay-time/network-delay-time.cpp
class Solution {
public:
    void buildGraph(vector<vector<pair<int, int>>> &adj, vector<vector<int>>& times)
    {
        // Iterate through times and populate adjacency list
        for (int i = 0; i < times.size(); i++) 
        {
            int u = times[i][0];
            int v = times[i][1];
            int t = times[i][2];

            adj[u].push_back({t, v});
        }
    }

    int networkDelayTime(vector<vector<int>>& times, int n, int src) 
    {
        // create adjacency list
        vector<vector<pair<int, int>>> adj(n + 1);
        buildGraph(adj, times);

        vector<int> time(n + 1, INT_MAX);
        time[src] = 0;

        //to optimise the dijaktras
        vector<bool> vis(n + 1, false);

        // min heap for pairs {time, node}
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        pq.push({0, src});

        while(!pq.empty())
        {
            auto p = pq.top();
            pq.pop();

            int currTime = p.first;
            int currNode = p.second;

            if(vis[currNode])
            continue;

            for(auto pair : adj[currNode])
            {
                int nbr = pair.second;
                int travelTime = pair.first;

                if(currTime + travelTime < time[nbr])
                {
                    time[nbr] = currTime + travelTime;
                    pq.push({time[nbr], nbr});
                }
            }

            vis[currNode] = true;
        }

        int maxTime = 0;
        for(int i = 1; i < time.size(); i++)
        maxTime = max(maxTime, time[i]);

        return maxTime == INT_MAX ? -1 : maxTime;
    }
};

File: 744-network-delay-time/network-delay-time.java
class Solution {
    private List<List<Integer>>[] buildGraph(int n, int[][] times){
        List<List<Integer>>[] adj = new ArrayList[n + 1];
        for (int i = 0; i <= n; i++) {
            adj[i] = new ArrayList<>();
        }

        for(int[] time : times) {
            int currNode = time[0];
            int neighbour = time[1];
            int distance = time[2];
            adj[currNode].add(List.of(neighbour, distance));
        }

        return adj;
    }

    public int networkDelayTime(int[][] times, int n, int k) {
        var adj = buildGraph(n, times);
        int[] dist = new int[n + 1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[k] = 0;
        Queue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        pq.offer(new int[]{0, k});

        while (!pq.isEmpty()) {
            var curr = pq.poll();
            int distance = curr[0];
            int node = curr[1];

            if (distance > dist[node]) continue;

            for(var adjNode : adj[node]) {
                int nbr = adjNode.get(0);
                int weight = adjNode.get(1);
                if(distance + weight < dist[nbr]) {
                    dist[nbr] = distance + weight;
                    pq.offer(new int[]{dist[nbr], nbr});
                }
            }
        }
        
        int minTime = 0;
        for(int i = 1; i <= n; i++) {
            if(dist[i] == Integer.MAX_VALUE)
            return -1;
            minTime = Math.max(minTime, dist[i]);
        }
        return minTime;
    }
}